server/src/main/java/org/elasticsearch/action/get/TransportGetFromTranslogAction.java

```diff
@@ -0,0 +1,187 @@
+                    getRequest.storedFields(),
+                    getRequest.realtime(),
+                    getRequest.version(),
+                    getRequest.versionType(),
+                    getRequest.fetchSourceContext(),
+                    getRequest.isForceSyntheticSource()
+                );
+            long segmentGeneration = -1;
+            if (result == null) {
+                segmentGeneration = ((InternalEngine) engine).getLastUnsafeSegmentGenerationForGets();
+            }
+            return new Response(result, segmentGeneration);
+        });
+    }
+
+    public static class Request extends ActionRequest {
+
+        private final GetRequest getRequest;
+        private final ShardId shardId;
+
+        public Request(GetRequest getRequest, ShardId shardId) {
+            this.getRequest = Objects.requireNonNull(getRequest);
+            this.shardId = Objects.requireNonNull(shardId);
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            getRequest = new GetRequest(in);
+            shardId = new ShardId(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            getRequest.writeTo(out);
+            shardId.writeTo(out);
+        }
+
+        public GetRequest getRequest() {
+            return getRequest;
+        }
+
+        public ShardId shardId() {
+            return shardId;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            return null;
+        }
+
+        @Override
+        public String toString() {
+            return "Request{" + "getRequest=" + getRequest + ", shardId=" + shardId + "}";
+        }
+    }
+
+    public static class Response extends ActionResponse {
+        private GetResult getResult;
+        /**
+         * The segment generation that the search shard should wait for before handling the real-time GET request locally.
+         * -1 if the result is not null (i.e., the result is served from the indexing shard), or there hasn't simply been
+         * any switches from unsafe to safe map in the LiveVersionMap (see {@link InternalEngine#getVersionFromMap(BytesRef)}).
+         */
+        private final long segmentGeneration;
+
+        public Response(GetResult getResult, long segmentGeneration) {
+            this.getResult = getResult;
+            this.segmentGeneration = segmentGeneration;
+        }
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            segmentGeneration = in.readZLong();
+            if (in.readBoolean()) {
+                getResult = new GetResult(in);
+            }
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeZLong(segmentGeneration);
+            if (getResult == null) {
+                out.writeBoolean(false);
```

Comment:
Do we really need to write a boolean? If segmentGeneration != -1, then the result is null and you simply do not need to write anything else.